tests/test_privateai.py

```diff
@@ -0,0 +1,82 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+from nemoguardrails import RailsConfig
+from tests.utils import TestChat
+
+
+@pytest.mark.unit
+def test_privateai_pii_detection_input_output():
+    config = RailsConfig.from_content(
+        yaml_content="""
+            models: []
+            rails:
+              config:
+                privateai:
+                  server_endpoint: https://dummy.private-ai.server/cloud/v3/process/text
+                  api_key: test_api_key
+                  input:
+                    entities:
+                      - EMAIL_ADDRESS
+                      - NAME
+                  output:
+                    entities:
+                      - EMAIL_ADDRESS
+                      - NAME
+              input:
+                flows:
+                  - detect pii on input
+              output:
+                flows:
+                  - detect pii on output
+        """,
+        colang_content="""
+            define user express greeting
+              "hi"
+
+            define flow
+              user express greeting
+              bot express greeting
+
+            define bot inform answer unknown
+              "I can't answer that."
+        """,
+    )
+
+    chat = TestChat(
+        config,
+        llm_completions=[
+            "  express greeting",
+            '  "Hi! My name is John as well."',
+        ],
+    )
+
+    def mock_detect_pii(return_value=True):
+        def mock_request(*args, **kwargs):
+            return return_value
+
+        return mock_request
+
+    chat.app.register_action(mock_detect_pii(True), "detect_pii")
+
+    # This will trigger the input rail
+    chat >> "Hi! I am Mr. John! And my email is test@gmail.com"
+    chat << "I can't answer that."
+
+    # This will trigger only the output one
+    chat >> "Hi!"
+    chat << "I can't answer that."
```

Comment:
- Please add tests when no rails get activated. 

- it is not readily clear which rail is blocking the generation.

- Please add tests for retrieval too, you can define following and register it as action

```python

@action()
def retrieve_relevant_chunks():
    context_updates = {"relevant_chunks": "Mock retrieved context."}

    return ActionResult(
        return_value=context_updates["relevant_chunks"],
        context_updates=context_updates,
    )

```

Have a look at [this test](https://github.com/NVIDIA/NeMo-Guardrails/blob/main/tests/test_patronus_lynx.py).

Reply:
> Please add tests for retrieval too, you can define following and register it as action

Testing retrieval, truly, seems a bit tricky. I feel we might need to modify the retrival rails or the [way I'm testing it](https://github.com/privateai/NeMo-Guardrails/blob/feature/private-ai-integration/tests/test_privateai.py#L166).



I was thinking of generating a bot utterance event if `$relevant_chunks` has a PII.

```
define subflow detect pii on retrieval
  """Check if the relevant chunks from the knowledge base have any PII."""
  $has_pii = execute detect_pii(source="retrieval", text=$relevant_chunks)

  if $has_pii
    bot inform answer unknown
    stop
```


But testing it with a mocked `retrieve_relevant_chunks`, make the process go in an infinite loop of `botIntent > .... > InternalSystemActionFinished > botIntent  > ....  > InternalSystemActionFinished` and end up with a `Exception: Too many events.`

If I understand it right, it's an intended behaviour and happening because my retreival flow always (in the test) ends up with a botIntent which in turn invoke a retreival flow.

Reply:
Updated the tests. Let me know what do you think about this.